stacks-signer/src/signerdb.rs

```diff
@@ -389,6 +403,71 @@ CREATE TABLE IF NOT EXISTS block_rejection_signer_addrs (
     PRIMARY KEY (signer_addr)
 ) STRICT;"#;
 
+// A lighter blocks table to aid in calculating tenure processing times
+// Will only track the most recent tenure blocks
+static CREATE_TENURE_BLOCKS_TABLE: &str = r#"
+CREATE TABLE IF NOT EXISTS tenure_blocks (
+    signer_signature_hash TEXT NOT NULL PRIMARY KEY,
+    reward_cycle INTEGER NOT NULL,
+    consensus_hash TEXT NOT NULL,
+    proposed_time INTEGER NOT NULL,
+    validation_time_ms INTEGER NOT NULL,
+    stacks_height INTEGER NOT NULL,
+    tenure_change INTEGER NOT NULL
+) STRICT;"#;
+
+// Migration logic necessary to move from blocks into tenure_blocks table
+// It will only migrate globally accepted blocks that are less than 2 reward cycles old.
+static MIGRATE_GLOBALLY_ACCEPTED_BLOCKS_TO_TENURE_BLOCKS: &str = r#"
+INSERT INTO tenure_blocks (
+    signer_signature_hash,
+    reward_cycle,
+    consensus_hash,
+    proposed_time,
+    validation_time_ms,
+    stacks_height,
+    tenure_change
+)
+SELECT 
+    signer_signature_hash,
+    reward_cycle,
+    consensus_hash,
+    json_extract(block_info, '$.proposed_time') AS proposed_time,
+    COALESCE(json_extract(block_info, '$.validation_time_ms'), 0) AS validation_time_ms,
+    stacks_height,
+    json_extract(block_info, '$.tenure_change') AS tenure_change
+FROM blocks
+WHERE json_extract(block_info, '$.state') = 'GloballyAccepted'
+  AND reward_cycle + 2 > (
+      SELECT MAX(reward_cycle) FROM blocks
+  );"#;
+
+static CREATE_TENURE_BLOCKS_ON_BLOCKS_TRIGGER: &str = r#"
+CREATE TRIGGER insert_into_tenure_blocks
```

Comment:
This seems wasteful without further insight or explanation.  Why is a separate table and trigger needed, instead of just indexing and querying the relevant fields in `blocks`?  It looks like there's only one place where you query the `tenure_blocks` table.

Reply:
This was in response to https://github.com/stacks-network/stacks-core/pull/5478#pullrequestreview-2446292269. This seemed like the simplest paired down solution that did not introduce any unnecessary edge cases. I do agree that it may be overengineering a potential non issue (originally thought it would be much easier to just cache a cumulative amount per tenure that reset per tenure change tx but this did not account for out of order blocks well and required a lot more heavy handed logic. This was what seemed to be cleanest solution that still gave some optimizations)

Reply:
I'm not sure how this could be that much more efficient than just querying the data directly from the `blocks` table?  I'm also not sure what exactly this table is providing that `blocks` isn't?

* This trigger unconditionally copies the contents of some rows and columns in `blocks` into `tenure_blocks` in insert order, so whatever is present in `tenure_blocks` for a given `signer_signature_hash` will also be in `blocks` [1].
* If this DB is instantiated in WAL mode (which it should be in any case), then reading from `blocks` won't ever block
* The code never inserts into `tenure_blocks` outside of this trigger

[1] I realize that the primary key on `blocks` is `(reward_cycle, signer_signature_hash)`, but in practice `signer_signature_hash` is globally unique anyway, so there shouldn't ever be two rows in `blocks` with the same `signer_signature_hash` but different reward cycles.

Reply:
Yeah, I'm debating the value of this change as well. The TLDR is that this PR will make writes slower and reads faster.

On the read side, the main difference is that we don't have to deserialize the entire block - that's a lot of bytes and also compute (relatively).

But, assuming we have the new `tenure_change` property, we can just do a `SELECT` with just the `block_info` properties we want. In that case, reads are a _tiny_ bit slower due to the `json_extract`, but pretty minor.

Reply:
If we're creating a new schema anyway, you could simply extract all of the relevant JSON fields and block header fields into their own columns preemptively, so you can avoid deserialization and `json_extract` on read (which would be a win no matter how you did it).  I _think_ you read a block's row more often than you write new blocks, so that would be a win.

Reply:
Done :D